File: src/java/foundgine-providers/src/main/java/com/foundgine/providers/storage/sql/SqlExecutionProvider.java
package com.foundgine.providers.storage.sql;

import com.foundgine.core.execution.*;
import com.foundgine.providers.storage.sql.query.CursorCodec;
import com.foundgine.providers.storage.sql.query.SqlParameterBinding;

import java.sql.*;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.*;

/**
 * Executes an already-compiled SQL plan against a JDBC connection.
 *
 * <p>The provider deliberately owns execution only: compilation remains in
 * {@link SqlCompiler}. Connection opening, caller transaction participation,
 * runtime authorization-context binding, cancellation and result materialization
 * happen here at the physical provider boundary.</p>
 */
public final class SqlExecutionProvider implements IExecutionProvider {
    private final Connection connection;
    private final Transaction transaction;
    private final boolean ownsTransaction;

    /** Optional statement hook retained for compatibility with existing callers. */
    public interface Transaction { void apply(Statement statement) throws SQLException; }

    /** Uses the caller's current JDBC transaction and does not commit/rollback it. */
    public SqlExecutionProvider(Connection connection) {
        this(connection, null, false);
    }

    /**
     * Compatibility constructor for callers that need to configure a statement
     * before execution. The provider does not infer transaction ownership from
     * this callback.
     */
    public SqlExecutionProvider(Connection connection, Transaction transaction) {
        this(connection, transaction, false);
    }

    /**
     * Explicit transaction-ownership constructor. When ownership is true and
     * the connection is currently in auto-commit mode, the provider starts and
     * owns a transaction for the execution. Existing caller transactions are
     * never committed or rolled back by this provider.
     */
    public SqlExecutionProvider(Connection connection, Transaction transaction, boolean ownsTransaction) {
        this.connection = Objects.requireNonNull(connection, "connection");
        this.transaction = transaction;
        this.ownsTransaction = ownsTransaction;
    }

    @Override
    public CompletionStage<ExecutionResult> executeAsync(ProviderPlan plan, ExecutionContext context, CancellationToken token) {
        Objects.requireNonNull(plan, "plan");
        Objects.requireNonNull(context, "context");
        Objects.requireNonNull(token, "token");
        if (!(plan instanceof SqlPlan sqlPlan)) {
            return CompletableFuture.failedFuture(new IllegalArgumentException("The SQL provider requires a SqlPlan."));
        }
        if (sqlPlan.authorizationBinding() == null) {
            return CompletableFuture.failedFuture(new IllegalStateException(
                    "The SQL provider refuses to execute a provider plan without authorization provenance."));
        }

        try {
            token.throwIfCancellationRequested();
            context.ensureWithinDeadline();
            long start = System.nanoTime();
            if (connection.isClosed()) {
                return CompletableFuture.failedFuture(new SQLException("The JDBC connection is closed."));
            }

            boolean originalAutoCommit = connection.getAutoCommit();
            boolean startedTransaction = false;
            if (ownsTransaction && originalAutoCommit) {
                connection.setAutoCommit(false);
                startedTransaction = true;
            }

            try {
                return executeStatement(sqlPlan, context, token, start, startedTransaction, originalAutoCommit);
            } catch (Throwable failure) {
                if (startedTransaction) rollbackQuietly();
                if (startedTransaction) restoreAutoCommit(originalAutoCommit);
                return CompletableFuture.failedFuture(failure);
            }
        } catch (Throwable t) {
            return CompletableFuture.failedFuture(t);
        }
    }

    private CompletionStage<ExecutionResult> executeStatement(
            SqlPlan sqlPlan,
            ExecutionContext context,
            CancellationToken token,
            long start,
            boolean startedTransaction,
            boolean originalAutoCommit) throws Exception {
        String jdbcSql = JdbcSqlPlaceholderRewriter.rewrite(sqlPlan.commandText());
        try (PreparedStatement statement = connection.prepareStatement(jdbcSql)) {
            if (transaction != null) transaction.apply(statement);

            AutoCloseable cancellationRegistration = token.register(() -> {
                try {
                    statement.cancel();
                } catch (SQLException ignored) {
                    // Best effort: cancellation remains observable through the token.
                }
            });

            try (cancellationRegistration) {
                bindParameters(statement, sqlPlan, context, token);
                token.throwIfCancellationRequested();
                try (ResultSet rs = statement.executeQuery()) {
                    List<ExecutionRow> rows = readRows(rs, sqlPlan, token);
                    ExecutionPageInfo pageInfo = buildPageInfo(rows, sqlPlan, context);
                    token.throwIfCancellationRequested();

                    if (startedTransaction) connection.commit();

                    long elapsed = Duration.ofNanos(System.nanoTime() - start).toMillis();
                    List<Integer> nodes = sqlPlan.authorization().stream()
                            .map(SqlAuthorizationPredicate::nodeId)
                            .toList();
                    String fingerprint = ExecutionEvidenceFactory.hash(buildPlanFingerprint(sqlPlan));
                    ExecutionEvidence evidence = ExecutionEvidenceFactory.create(
                            "sql", fingerprint, nodes, rows.size(), elapsed, sqlPlan.commandText());
                    return CompletableFuture.completedFuture(new ExecutionResult(rows, pageInfo, evidence, null));
                }
            } finally {
                if (startedTransaction) restoreAutoCommit(originalAutoCommit);
            }
        } catch (Throwable failure) {
            if (startedTransaction) rollbackQuietly();
            if (startedTransaction) restoreAutoCommit(originalAutoCommit);
            throw failure;
        }
    }

    private static void bindParameters(PreparedStatement statement, SqlPlan plan,
                                       ExecutionContext context, CancellationToken token) throws SQLException {
        int index = 1;
        for (SqlParameterBinding binding : plan.effectiveParameters()) {
            token.throwIfCancellationRequested();
            Object value = binding.value();
            if (binding.contextPath() != null) {
                if (!context.effectiveValues().containsKey(binding.contextPath())) {
                    throw new IllegalStateException("Execution context does not contain authorization value '" +
                            binding.contextPath() + "'.");
                }
                value = context.effectiveValues().get(binding.contextPath());
                if (ExecutionContextKeys.PAGINATION_LIMIT.equals(binding.contextPath()) &&
                        plan.pagination() != null && value instanceof Number n) {
                    value = n.intValue() + 1;
                }
            }
            if (value == null) statement.setObject(index, null);
            else statement.setObject(index, value);
            index++;
        }
    }

    private static List<ExecutionRow> readRows(ResultSet rs, SqlPlan plan, CancellationToken token) throws SQLException {
        List<ExecutionRow> rows = new ArrayList<>();
        Map<String, SqlColumnBinding> bindings = new HashMap<>();
        for (SqlColumnBinding b : plan.columns()) bindings.put(b.resultName(), b);
        ResultSetMetaData meta = rs.getMetaData();
        while (rs.next()) {
            token.throwIfCancellationRequested();
            Map<String, Object> values = new LinkedHashMap<>();
            Map<ExecutionCellKey, Object> cells = new LinkedHashMap<>();
            for (int i = 1; i <= meta.getColumnCount(); i++) {
                String name = meta.getColumnLabel(i);
                if (name == null || name.isEmpty()) name = meta.getColumnName(i);
                Object value = rs.getObject(i);
                values.put(name, value);
                SqlColumnBinding b = bindings.get(name);
                if (b != null) {
                    cells.put(new ExecutionCellKey(b.nodeId(), b.entityId(), b.fieldId()), value);
                }
            }
            rows.add(new ExecutionRow(values, cells));
        }
        return rows;
    }

    private static ExecutionPageInfo buildPageInfo(List<ExecutionRow> rows, SqlPlan plan, ExecutionContext context) {
        if (plan.pagination() == null) return null;
        var paging = plan.pagination();
        int first = paging.first();
        Object limit = context.effectiveValues().get(ExecutionContextKeys.PAGINATION_LIMIT);
        if (limit instanceof Number n) first = n.intValue();
        boolean hasCursor = paging.after() != null;
        Object cursor = context.effectiveValues().get(ExecutionContextKeys.PAGINATION_HAS_CURSOR);
        if (cursor instanceof Boolean b) hasCursor = b;

        boolean hasNext = rows.size() > first;
        if (hasNext) rows.remove(rows.size() - 1);
        if (rows.isEmpty()) return new ExecutionPageInfo(null, null, hasNext, hasCursor);

        String start = encodeCursor(rows.get(0), paging.cursorValues());
        String end = encodeCursor(rows.get(rows.size() - 1), paging.cursorValues());
        return new ExecutionPageInfo(start, end, hasNext, hasCursor);
    }

    private static String encodeCursor(ExecutionRow row, List<SqlCursorBinding> bindings) {
        List<Object> values = new ArrayList<>();
        for (var b : bindings) {
            Object value = row.values().get(b.resultName());
            if (value == null) {
                throw new IllegalStateException("Cursor ordering fields cannot contain null values.");
            }
            values.add(value);
        }
        return CursorCodec.encode(values);
    }

    private static String buildPlanFingerprint(SqlPlan p) {
        String cols = p.columns().stream()
                .map(x -> x.resultName() + "|" + x.entityId().value() + "|" + x.fieldId().value() +
                        "|" + x.columnName() + "|" + x.nodeId() + "|" + x.isCursor())
                .sorted().reduce("", (a, b) -> a + ";" + b);
        String auth = p.authorization().stream()
                .map(x -> x.nodeId() + "|" + x.predicate())
                .sorted().reduce("", (a, b) -> a + ";" + b);
        return p.commandText() + "|" + cols + "|" + auth;
    }

    private void rollbackQuietly() {
        try { connection.rollback(); } catch (SQLException ignored) { }
    }

    private void restoreAutoCommit(boolean original) {
        try { connection.setAutoCommit(original); } catch (SQLException ignored) { }
    }
}

File: src/csharp/Foundgine.Providers/Storage/Sql/SqlExecutionProvider.cs
using Foundgine.Core.Abstractions;
using System.Data.Common;
using System.Diagnostics;
using Foundgine.Core.Execution;
using ExecutionContext = Foundgine.Core.Execution.ExecutionContext;

namespace Foundgine.Providers.Storage.Sql;

/// <summary>
/// Executes an already-compiled SQL plan against an ADO.NET connection.
/// Compilation and execution remain separate responsibilities.
/// </summary>
public sealed class SqlExecutionProvider : IExecutionProvider
{
    private readonly DbConnection _connection;
    private readonly DbTransaction? _transaction;

    public SqlExecutionProvider(DbConnection connection, DbTransaction? transaction = null)
    {
        _connection = connection ?? throw new ArgumentNullException(nameof(connection));
        _transaction = transaction;
    }

    public async Task<ExecutionResult> ExecuteAsync(
        ProviderPlan plan,
        ExecutionContext context,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(plan);
        if (plan is not SqlPlan sqlPlan)
            throw new ArgumentException("The SQL provider requires a SqlPlan.", nameof(plan));
        if (sqlPlan.AuthorizationBinding is null)
            throw new InvalidOperationException(
                "The SQL provider refuses to execute a provider plan without authorization provenance.");

        var stopwatch = Stopwatch.StartNew();

        if (_connection.State != System.Data.ConnectionState.Open)
            await _connection.OpenAsync(cancellationToken);

        await using var command = _connection.CreateCommand();
        command.Transaction = _transaction;
        command.CommandText = sqlPlan.CommandText;
        foreach (var binding in sqlPlan.EffectiveParameters)
        {
            var parameter = command.CreateParameter();
            parameter.ParameterName = "@" + binding.Name;
            object? value = binding.Value;
            if (binding.ContextPath is { } contextPath)
            {
                if (!context.TryGetValue(contextPath, out value))
                    throw new InvalidOperationException(
                        $"Execution context does not contain authorization value '{contextPath}'.");

                // Forward pagination fetches one extra row so the provider can
                // determine HasNextPage without changing the requested page size.
                if (string.Equals(contextPath, ExecutionContextKeys.PaginationLimit, StringComparison.Ordinal) &&
                    sqlPlan.Pagination is not null &&
                    value is not null)
                {
                    value = Convert.ToInt32(value, System.Globalization.CultureInfo.InvariantCulture) + 1;
                }
            }

            parameter.Value = value ?? DBNull.Value;
            command.Parameters.Add(parameter);
        }

        await using var reader = await command.ExecuteReaderAsync(cancellationToken);
        var rows = new List<ExecutionRow>();

        while (await reader.ReadAsync(cancellationToken))
        {
            var values = new Dictionary<string, object?>(StringComparer.Ordinal);
            var cells = new Dictionary<ExecutionCellKey, object?>();

            for (var i = 0; i < reader.FieldCount; i++)
            {
                var name = reader.GetName(i);
                var value = reader.IsDBNull(i) ? null : reader.GetValue(i);
                values[name] = value;

                var binding = sqlPlan.Columns.FirstOrDefault(x =>
                    string.Equals(x.ResultName, name, StringComparison.Ordinal));

                if (binding is not null)
                {
                    cells[new ExecutionCellKey(
                        binding.NodeId,
                        binding.EntityId,
                        binding.FieldId)] = value;
                }
            }

            rows.Add(new ExecutionRow(values, cells));
        }

        ExecutionPageInfo? pageInfo = null;
        if (sqlPlan.Pagination is { } paging)
        {
            var first = paging.First;
            var hasCursor = paging.After is not null;
            if (context.TryGetValue(ExecutionContextKeys.PaginationLimit, out var runtimeLimit) &&
                runtimeLimit is not null)
            {
                first = Convert.ToInt32(runtimeLimit, System.Globalization.CultureInfo.InvariantCulture);
            }

            if (context.TryGetValue(ExecutionContextKeys.PaginationHasCursor, out var cursorValue) &&
                cursorValue is bool runtimeCursor)
                hasCursor = runtimeCursor;

            var hasNext = rows.Count > first;
            if (hasNext) rows.RemoveAt(rows.Count - 1);
            string? start = null;
            string? end = null;
            if (rows.Count > 0)
            {
                var firstValues = paging.CursorValues
                    .Select(binding => rows[0].Values[binding.ResultName])
                    .ToArray();
                var lastValues = paging.CursorValues
                    .Select(binding => rows[^1].Values[binding.ResultName])
                    .ToArray();

                if (firstValues.Any(value => value is null) || lastValues.Any(value => value is null))
                    throw new InvalidOperationException("Cursor ordering fields cannot contain null values.");

                start = Query.CursorCodec.Encode(firstValues);
                end = Query.CursorCodec.Encode(lastValues);
            }

            pageInfo = new ExecutionPageInfo(start, end, hasNext, hasCursor);
        }

        stopwatch.Stop();
        var evidence = ExecutionEvidenceFactory.Create(
            "sql",
            ExecutionEvidenceFactory.Hash(BuildPlanFingerprint(sqlPlan)),
            sqlPlan.Authorization?.Select(x => x.NodeId) ?? [],
            rows.Count,
            stopwatch.ElapsedMilliseconds,
            sqlPlan.CommandText);

        return new ExecutionResult(rows, pageInfo, evidence);
    }

    private static string BuildPlanFingerprint(SqlPlan plan)
    {
        var columns = string.Join(";", plan.Columns.Select(x =>
            $"{x.ResultName}|{x.EntityId.Value}|{x.FieldId.Value}|{x.ColumnName}|{x.NodeId}|{x.IsCursor}"));
        var authorization = string.Join(";", plan.Authorization?.Select(x =>
            $"{x.NodeId}|{x.Predicate}") ?? []);
        return $"{plan.CommandText}|{columns}|{authorization}";
    }
}